lib/Frontend/nvqpp/ConvertDecl.cpp

```diff
@@ -153,66 +120,351 @@ QuakeBridgeVisitor::getOrAddFunc(Location loc, StringRef funcName,
   if (func) {
     if (!func.empty()) {
       // Already lowered function func, skip it.
-      return {func, true};
+      return {func, /*defined=*/true};
     }
-  } else {
-    OpBuilder build(module.getBodyRegion());
-    OpBuilder::InsertionGuard guard(build);
-    build.setInsertionPointToEnd(module.getBody());
-    SmallVector<NamedAttribute> attrs;
-    func = build.create<func::FuncOp>(loc, funcName, funcTy, attrs);
-    func.setPrivate();
+    // Function was declared but not defined.
+    return {func, /*defined=*/false};
   }
-  return {func, false};
+  // Function not found, so add it to the module.
+  OpBuilder build(module.getBodyRegion());
+  OpBuilder::InsertionGuard guard(build);
+  build.setInsertionPointToEnd(module.getBody());
+  SmallVector<NamedAttribute> attrs;
+  func = build.create<func::FuncOp>(loc, funcName, funcTy, attrs);
+  func.setPrivate();
+  return {func, /*defined=*/false};
 }
 
-bool QuakeBridgeVisitor::VisitFunctionDecl(clang::FunctionDecl *x) {
-  if (typeMode)
+bool QuakeBridgeVisitor::interceptRecordDecl(clang::RecordDecl *x) {
+  // Some decls will be intercepted and replaced with high-level types in quake.
+  // Do this here to avoid traversing their fields, etc.
+  auto *ident = x->getIdentifier();
+  if (!ident || x->isLambda())
+    return false;
+  auto name = ident->getName();
+  auto *ctx = builder.getContext();
+  if (isInNamespace(x, "cudaq")) {
+    // Types from the `cudaq` namespace.
+    // A qubit is a qudit<LEVEL=2>.
+    if (name.equals("qudit") || name.equals("qubit"))
+      return pushType(quake::RefType::get(ctx));
+    // qreg<SIZE,LEVEL>, qarray<SIZE,LEVEL>, qspan<SIZE,LEVEL>
+    if (name.equals("qspan") || name.equals("qreg") || name.equals("qarray")) {
+      // If the first template argument is not `std::dynamic_extent` then we
+      // have a constant sized VeqType.
+      if (auto *tempSpec =
+              dyn_cast<clang::ClassTemplateSpecializationDecl>(x)) {
+        auto templArg = tempSpec->getTemplateArgs()[0];
+        assert(templArg.getKind() ==
+               clang::TemplateArgument::ArgKind::Integral);
+        std::int64_t size = templArg.getAsIntegral().getExtValue();
+        if (size != static_cast<std::int64_t>(std::dynamic_extent))
+          return pushType(quake::VeqType::get(ctx, size));
+      }
+      return pushType(quake::VeqType::getUnsized(ctx));
+    }
+    // qvector<LEVEL>, qview<LEVEL>
+    if (name.equals("qvector") || name.equals("qview"))
+      return pushType(quake::VeqType::getUnsized(ctx));
+    if (name.equals("ctrl") /*|| name.equals("adj") || name.equals("base")*/)
+      return pushType({});
```

Comment:
Will this cause issues for anything popping off the last type?

Reply:
This is a NOP to meet the invariant here. If the type is intercepted (meaning the bridge is filtering out this C++ type and replacing it in whole with a Quake type, ignoring the C++ definition), then the invariant is that a `mlir::Type` is pushed on the top of the stack. For `cudaq::ctrl` that Quake type is very uninteresting and technically-speaking erased. So this null Type will be popped and the code will generate an mlir attribute.

On the other hand, I think a case could be made that nothing should be pushed on the type stack here, as the other two aren't doing so.

Bottom line: I need to investigate this more.